[Feat/#13] SNS 관련 로직 구현

sung-silver opened the pull request:
<!--
name: Makers Push Notification Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closed #13 

## Work Description ✏️
1. SNS 실패 콜백 핸들러 (SnsHandler)
- SNSEvent 수신 및 처리
- 이벤트 유효성 검사
2. 토큰 추출 및 조회
- SNS 메시지에서 Token 추출 (extractTokenFromRecord)
- 토큰으로 DeviceTokenEntity 조회 (collectTokenEntities)
- DeviceTokenRepository.findByPk 구현
3. 실패 로그 기록
- createFailureHistory 메서드 구현
- HistoryEntity에 실패 상태, userId, messageId 저장
- DynamoDB에 저장 (HistoryRepository.save)
4. 토큰 무효화
- UserService.unregisterToken 구현
- SNS 구독 해제 (snsService.unsubscribe)
- SNS 엔드포인트 삭제 (snsService.cancelEndpoint)
- DynamoDB에서 UserEntity와 DeviceTokenEntity 삭제

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

rdd9223 commented on src/services/userService.ts:
```diff
@@ -95,8 +96,9 @@ const getUserByTokenId = async (deviceToken: string): Promise<DeviceTokenEntity
 const findTokenByUserIds = async (userIds: string[]): Promise<UserTokenEntity[]> => {
-  const result = await Promise.all(userIds.map(async (userId) => getTokenByUserId(userId)));
-  return result.filter((user: UserTokenEntity | null): user is UserTokenEntity => user !== null);
+  const tokens = await Promise.all(userIds.flatMap(async (userId) => getTokenByUserId(userId)));
```
요거는 flatmap 안해도 되지 않을까요...?

dragontaek-lee replied:
오잉 map으로 했었는데 반영이 안되었군요! 집어주셔서 감사합니다~!

dudgns3tp reviewed (commented):
영우가 잘 리뷰 해주었네요..! 반영 하시면 어프루브 해두겠습니다~

dudgns3tp commented:
고생하셨습니다 🥹👍

dragontaek-lee commented on src/services/userService.ts:
```diff
@@ -37,32 +37,33 @@ function isTokenUserEntity(queryCommandOutputItems: Record<string, AttributeValu
 }
 
-const getTokenByUserId = async (userId: string): Promise<UserTokenEntity | null> => {
+const getTokenByUserId = async (userId: string): Promise<UserTokenEntity[] | []> => {
```
아 그렇네요!

dudgns3tp reviewed (approved):
LGTM

sung-silver committed:
[FEAT] SnsService 구현

- subscribe, publish, endpoint 관리 로직 구현
- SNS publish 시 AWSServiceException, SdkClientException, RuntimeException으로 구분하여 예외처리

sung-silver committed:
[FEAT] NotificationService 구현

- 플랫폼별/전체 토픽 푸시 발송 로직 구현
- 공통 sendPushMessage 메서드로 SNS publish

sung-silver committed:
[FEAT] UserService에 토큰 무효화 로직 구현

- SNS 구독 해제
- SNS 엔드포인트 삭제
- DynamoDB 토큰 레코드 삭제

sung-silver committed:
[REFACTOR] ExternalException 주석 삭제

sung-silver committed:
[REFACTOR] if 조건문 boolean 변수로 리팩토링

sung-silver committed:
[FEAT] DeviceTokenRepository에 findByPk 메서드 구현

sung-silver committed:
[FEAT] SnsHandler 구현

- SNS 실패 콜백 핸들러 구현: SNS 실패 이벤트 수신 및 처리 + 실패한 푸시 알림에 대한 토큰 무효화 및 실패 로그 기록
- 토큰 추출 및 조회 로직: 토큰으로 DeviceTokenEntity 조회, pk로 토큰 엔티티 Query
- 실페 레코드 처리: 실패 로그 저장

sung-silver committed:
[FEAT] UserService에 registerToken 구현

sung-silver committed:
[REFACTOR] SnsService에 json 문자열 상수화

sung-silver committed:
[REFACTOR] SnsService의 application protocol 상수화

sung-silver committed:
[CHORE] 코드 포매팅
